pyroomacoustics/open_sofa_interpolate.py

```diff
@@ -0,0 +1,556 @@
+    res = np.insert(res, len(res), np.pi)
+    res = np.insert(res, 0, 0)
+
+    w = -np.diff(np.cos(res))
+
+    w_ = np.tile(w, no_of_lat) #Repeat matrice n times
+
+    w_ = np.diag(w_) #Diagnol of the matrice (no_of_nodes * no_of_nodes)
+
+    Ysh_tilda = np.matmul(w_, Ysh)
+
+    Ysh_tilda_inv = np.linalg.pinv(Ysh_tilda, rcond=1e-2) #rcond is inverse of the condition number
+
+    print("Condition Number Psuedo Inverse ",np.linalg.cond(Ysh_tilda_inv),Ysh_tilda_inv.shape)
+
+    return Ysh_tilda_inv,w_
+
+
+
+
+def cart2sphere(points):
+    # Convert cartesian coordinates into spherical coordinates (radians)
+
+    r = np.sqrt(np.square(points[:, 0]) + np.square(points[:, 1]) + np.square(points[:, 2]))
+
+    theta_fibo = np.arccos((points[:, 2] / r))
+
+    phi_fibo = np.arctan2(points[:, 1], points[:, 0])
+
+
+    #phi_fibo += np.pi  # phi_fibo was in range of [-np.pi,np.pi]
+
+    return phi_fibo,theta_fibo,r
+
+def sph2cart(phi,theta,r):
+    #Convert spherical cordinates to cartesian cordinates.
+
+    x=np.cos(phi)*np.sin(theta)
+    y=np.sin(phi)*np.sin(theta)
+    z=np.cos(theta)
+
+    return x,y,z
+
+
+class DIRPATInterpolate():
+
+    """
+    Intialization class performs the following functions
+    Open DIRPAT Files
+    Interpolate on fibonacci sphere
+    Rotate the pattern w.r.t to the given azimuth(phi) [-pi, pi] and colatitude(theta) [0,pi] defines the orientation of the source or a receiver
+    Reutrns response in the given direction w.r.t to the given angle of arrivals(microphones) / angle of departures(angle of departures)
+
+    Parameters
+    --------------
+    path (string) : Path towards the specific DIRPAT file
+    fs (int) : Sampling frequency of the microphone or source FIR's filters , should be  <= simulation frequency
+    directivity_pattern (int) : The specific pattern in the DIRPAT files are associated with id's , presented in the github document
+    source (Boolean) : If the DIRPAT files represnts one of source or receiver's
+    interpolate (Boolean) : Interpolate the FIR filter or not
+    no_of_points_fibo_sphere (int) : How many points should be there on fibonnacci sphere
+    azimuth_ (int) , colatitude_ : Orientation of the source/mic in the simulation environment.
+
+
+    """
+
+    def __init__(self,path,fs,directivity_pattern=2,source=False,interpolate=True,no_of_points_fibo_sphere=1200,azimuth_simulation=0,colatitude_simulation=0):
+
+        self.path=path
+        self.id = directivity_pattern
+        self.source = source
+
+
+
+        self.fs=fs
+
+        self.phi_sofa, self.theta_sofa, self.r_sofa, self.theta_16, self.sofa_msr_fir, self.no_of_nodes, self.samples_size_ir, self.no_of_lat = self.open_sofa_database(path=self.path,
+            fs=self.fs)
+
+        self.sofa_x,self.sofa_y,self.sofa_z=sph2cart(self.phi_sofa,self.theta_sofa,self.r_sofa)
+
+        self.interpolate = interpolate
+
+        #Rotation matrix
+
+        n_c = (colatitude_simulation)
```

Comment:
The rotation here assumes that the original response is pointing to [1, 0, 0]. This is fine, but should be documented in the docstring.

Also, there is no rotation wrt the x-axis. This is fine for symmetric responses, such as cardioid, but probably not for measured ones. To be general, we probably want to provide a full rotation with 3 angles.